feat(service-worker): Support "Request"-Action

angelaki opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?

Notifications (and their actions) can set a Url that is simple fetched on click. This can be used for sending an answer to your application's API.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

angelaki committed:
feat(service-worker): support `request` as a `notificationclick` action

Implement a new `notificationclick` action, `request`, which will send a
GET request to the specified URL,,without opening a new window. This can
be useful for hitting a API endpoint.

angelaki commented:
Guess that'd be a pretty neat feature. Pushing welcome!

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -380,6 +380,10 @@ export class Driver implements Debuggable, UpdateSource {
         }
         break;
       }
+      case 'fetch' {
```
Maybe rename this to `request` (as it sounds more generic and we might want to expand it in the future)?

angelaki replied:
Yeah, sounds good! Was thinking about implementing other request types etc. in future. So quite good idea.

gkalpak commented on aio/content/guide/service-worker-notifications.md:
```diff
@@ -61,6 +61,7 @@ The Angular service worker supports the following operations:
 | `openWindow`                | Opens a new tab at the specified URL, which is resolved relative to the service worker scope.                                                                                                            |
 | `focusLastFocusedOrOpen`    | Focuses the last focused client. If there is no client open, then it opens a new tab at the specified URL, which is resolved relative to the service worker scope.                                       |
 | `navigateLastFocusedOrOpen` | Focuses the last focused client and navigates it to the specified URL, which is resolved relative to the service worker scope. If there is no client open, then it opens a new tab at the specified URL. |
+| `fetch` | Do nothing but sending a simple Http-GET to the specified URL. |
```
Nit:

```suggestion
| `fetch`                     | Send a simple GET request to the specified URL.                                                                                                                                                          |
```

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -380,6 +380,10 @@ export class Driver implements Debuggable, UpdateSource {
         break;
       }
+      case 'fetch' {
+        await fetch(urlToOpen);
```
This should access `fetch` through `scope` (to make testing easier):

```suggestion
        await this.scope.fetch(urlToOpen);
```

gkalpak reviewed (changes requested):
Thx, @angelaki! Looks like a reasonable and straight forward feature.
Would you be able to add some tests for this feature in the [happy_spec.ts][1] file?

[1]: https://github.com/angular/angular/blob/11285bfd472cbfba938aadc0fee7f68d1e01e6ba/packages/service-worker/worker/test/happy_spec.ts#L828

angelaki commented:
> Thx, @angelaki! Looks like a reasonable and straight forward feature. Would you be able to add some tests for this feature in the [happy_spec.ts](https://github.com/angular/angular/blob/11285bfd472cbfba938aadc0fee7f68d1e01e6ba/packages/service-worker/worker/test/happy_spec.ts#L828) file?

Sorry, I'm not super familiar with this. What could this test look like? (How) am I able to check if a request was called by the service worker?

gkalpak committed:
test(service-worker): remove redundant ternary operator from `MockRequest` initialization

Since the `init` parameter has a default value of `{}`, it can never be
`!== undefined`. Thus, it is not necessary to account for that case.

gkalpak committed:
fixup! feat(service-worker): support `request` as a `notificationclick` action

gkalpak committed:
docs(service-worker): remove duplication when describing `notificationclick` operations

gkalpak commented:
Thx for making the changes, @angelaki.

I pushed a few extra changes, including:
- Squashing commits.
- Updating the commit message to give more details on the implemented feature.
- Add tests for the new action.
- Fix a couple minor things (not directly related to this PR) that I noticed while working on the other fixes.

Unfortunately, CI seemed to be tripped off by the fact that this PR was targeting the `14.1.x` branch (instead of `main`).

Therefore, I have created PR #46912 with the same changes. Can you please take a look and let's continue the discussion there.

angelaki commented:
Thank you! When is this likely to be deployed? Will it make it to the next 14.1 build or go to 14.2?

gkalpak commented:
It will be released with the next minor after the PR is merged. So, it depends on how fast the PR is reviewed and merged :smiley:
